include/multi_blas_helper.cuh

```diff
@@ -1,23 +1,94 @@
+#pragma once
+
 #include <algorithm>
 #include <register_traits.h>
 #include <blas_helper.cuh>
+#include <kernel_helper.h>
+#include <target_device.h>
 
 namespace quda
 {
 
   namespace blas
   {
 
-    // storage for matrix coefficients
-#define MAX_MATRIX_SIZE 8192
-#define MAX_ARG_SIZE 4096
-    __constant__ signed char Amatrix_d[MAX_MATRIX_SIZE];
-    __constant__ signed char Bmatrix_d[MAX_MATRIX_SIZE];
-    __constant__ signed char Cmatrix_d[MAX_MATRIX_SIZE];
+    /**
+       @brief Returns the maximum size of a coefficient array for 2-d multi-blas
+    */
+    constexpr size_t max_array_size() { return 8192; }
+
+    /**
+       @brief Returns the maximum size of the kernel argument
+       @tparam Functor reducer and multi_1d blas kernels use kernel arg, others use constant
+    */
+    template <typename Functor>
+    constexpr size_t max_arg_size() { return (Functor::multi_1d || Functor::reducer) ? device::max_kernel_arg_size() : device::max_constant_size(); }
+
+    template <bool multi_1d = false, typename Arg, typename T> std::enable_if_t<multi_1d, void>
```

Comment:
Maybe I'm misunderstanding how `set_param` is used here, but should there be a comparison of `std::max(arg.NXZ, arg.NYW)` against `max_arg_size()`? (Analogous to the comparison against `max_array_size()` in the other instantiation of `set_param`, which makes sense to me.) Or is the theoretical constraint of `max_arg_size()` automatically handled by the constant memory/etc abstraction?

Reply:
I'm not fully understanding this comment, but let me explain a few points:
* `max_arg_size` is determined on line 25, and its size depends on whether we are doing a multi-1d/reducer or multi-2d
* If we're doing multi-1d/reducer then we always use the kernel arg path, and so the max arg size is set to 4 KiB on CUDA
* If we're doing multi-2d then we use the constant arg path, so the max arg size is the max constant size, which is presently set to 32 KiB on CUDA

Reply:
Rereading your question, maybe I get it now: there's no need for an extra check with the first `set_param` instance against the `max_arg_size` since if we exceed the size, we'd get a compile error from instantiating a kernel template with too great an arg size, and we'd also trip the `static_assert` in include/targets/cuda/tunable_kernel.h.

Reply:
Ah, okay, this set of explanations fully addresses my questions, thank you.